### PR TITLE
ESMDA

### LINKED ISSUE #73: Consider adding `ESMDA`
Could I think have a similar api to `SIES` but where we pass in a list of weights.

### PR DESCRIPTION
**Pull request by tommyod**
Closes #73

**Work done**

- Implemented the ESMDA algorithm from the 2012 and 2013 Emerick papers (full references in code).
- Implemented 7 different ways of solving the core numerical routine, the inversion problem $C_{MD} (C_{DD} + \alpha C_{D})^{-1}(D - Y)$. These approaches are tested against each other in 1000+ parameterized tests. They are also timed, see below.
- The implementation allows $C_{D}$ to be a 2D covariance matrix, or a diagonal 1D covariance matrix. Every inversion method exploits 1D $C_{D}$ for speed gains.
- Added `ensemble_mask` argument, allowing only a subset of ensemble members to go into computations.
- The numerical routines are quite tuned, e.g., I used blas routines from [scipy.linalg.blas](https://docs.scipy.org/doc/scipy/reference/linalg.blas.html) to perform some of the matrix operations. There is still some work to be done with respect to tuning in the future.

**API design**

Here is what the API currently looks like:

```python3
esmda = ESMDA(C_D, # Covariance matrix
              observations, # Vector of observations
              alpha=10, # Number of iterations (or vector or alphas)
              inversion="subspace", 
              seed=1)

for iteration in range(esmda.num_assimilations()):
    X_i = assimilate(X=X_i, Y=g(X_i)) # Assimilate data
```

**Discussion**

- I would like some feedback on the API. I implemented something like what I thought about in #111 .
- Other general comments are also very welcome.
- I am not entirely finished, but ready for some comments if someone has the time.

**Timing of inversion methods**

```
>>> test_timing(num_outputs=10_000, num_inputs=1000, ensemble_members=100)
--------------------------------
Function: inversion_exact_naive on dense covariance: 29.2188 s
Function: inversion_exact_naive on diagonal covariance: 27.2416 s
--------------------------------
Function: inversion_exact_cholesky on dense covariance: 5.0232 s
Function: inversion_exact_cholesky on diagonal covariance: 4.314 s
--------------------------------
Function: inversion_subspace on dense covariance: 0.6023 s
Function: inversion_subspace on diagonal covariance: 0.1252 s
--------------------------------
Function: inversion_exact_subspace_woodbury on dense covariance: 29.2699 s
Function: inversion_exact_subspace_woodbury on diagonal covariance: 1.8164 s
--------------------------------
Function: inversion_rescaled_subspace on dense covariance: 7.4716 s
Function: inversion_rescaled_subspace on diagonal covariance: 0.1046 s
--------------------------------
```

**Timing vs pyESMDA**

The implementation is significantly faster than [pyESMDA](https://pypi.org/project/pyesmda/), and competitive with ES / IES:

```
num_outputs = 5000
num_emsemble = 50
num_inputs = 500

ESMDA   - elapsed time: 0.9504 s
ES      - elapsed time: 1.0938 s
pyESMDA - elapsed time: 126.01 s
```

### CODE REVIEW

**Comment by tommyod** [comment]
Converted back to draft. Found some issues. Will look at it before review ...

File: src/iterative_ensemble_smoother/esmda.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+        # X_posterior = X_current + C_MD @ K
+        # K := sp.linalg.inv(C_DD + C_D_alpha) @ (D - Y)
+
+        # In the typical case where num_outputs >> num_inputs >> ensemble members,
+        # multiplying in the order below from the right to the left, i.e.,
+        # C_MD @ (inv(C_DD + alpha * C_D) @ (D - Y))
+        # is faster than the alternative order
+        # (C_MD @ inv(C_DD + alpha * C_D)) @ (D - Y)
+        X_posterior = np.copy(X)
```
Comment (dafeda) [inline src/iterative_ensemble_smoother/esmda.py:217]: We should aim for an implementation that does not make a copy of X as it can be large.

Comment (tommyod) [reply src/iterative_ensemble_smoother/esmda.py:228]: good point!

**Comment by tommyod** [comment]
@dafeda @Blunde1 - I am ready for comments :)

File: src/iterative_ensemble_smoother/esmda.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        "exact": inversion_exact_cholesky,
+        "subspace": inversion_subspace,
+    }
+
+    def __init__(
+        self,
+        C_D: npt.NDArray[np.double],
+        observations: npt.NDArray[np.double],
+        alpha: Union[int, npt.NDArray[np.double]] = 5,
+        seed: Optional[int] = None,
```
Comment (dafeda) [inline src/iterative_ensemble_smoother/esmda.py:52]: Nitpick, but I guess this should be typed as a union of int and generator.

File: src/iterative_ensemble_smoother/esmda.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+        -------
+        X_posterior : np.ndarray
+            2D array of shape (num_inputs, num_ensemble_members).
+
+        """
+        if self.iteration >= self.num_assimilations():
+            raise Exception("No more assimilation steps to run.")
+
+        # Verify shapes
+        num_inputs, num_ensemble = X.shape
```
Comment (dafeda) [inline src/iterative_ensemble_smoother/esmda.py:182]: `num_inputs` is not used

Comment (tommyod) [reply src/iterative_ensemble_smoother/esmda.py]: nice find. i ran [vulture](https://pypi.org/project/vulture/) and removed this, as well as some other, unused variables

File: src/iterative_ensemble_smoother/esmda.py (hunk 2 above)
Comment (tommyod) [reply src/iterative_ensemble_smoother/esmda.py]: True! fixed!

File: src/iterative_ensemble_smoother/esmda.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
+"""
+Ensemble Smoother with Multiple Data Assimilation (ES-MDA)
+----------------------------------------------------------
+
+Imlementation of the 2013 paper "Ensemble smoother with multiple data assimilation"
```
Comment (dafeda) [inline src/iterative_ensemble_smoother/esmda.py:5]: Typo: Should be "Implementation"

File: src/iterative_ensemble_smoother/esmda.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+            Covariance matrix of outputs of shape (num_outputs, num_outputs).
+            If a 1D array is passed, it represents a diagonal covariance matrix.
+        observations : np.ndarray
+            1D array of shape (num_inputs,) representing real-world observations.
+        alpha : int or 1D np.ndarray, optional
+            If an integer `alpha` is given, an array with length `alpha` and
+            elements `alpha` is constructed. If an 1D array is given, it is
+            normalized so sum_i 1/alpha_i = 1 and used. The default is 5, which
+            corresponds to np.array([5, 5, 5, 5, 5]).
+        seed : integer, optional
```
Comment (dafeda) [inline src/iterative_ensemble_smoother/esmda.py:71]: Type does not correspond to type-hint which is why I in general don't like documenting type-hints.

File: src/iterative_ensemble_smoother/esmda.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+        if not (
+            (isinstance(alpha, np.ndarray) and alpha.ndim == 1)
+            or isinstance(alpha, numbers.Integral)
+        ):
+            raise TypeError("Argument `alpha` must be an integer or a 1D NumPy array.")
+
+        if not (
+            isinstance(seed, (int, np.random._generator.Generator)) or seed is None
+        ):
+            raise TypeError("Argument `seed` must be an integer.")
```
Comment (dafeda) [inline src/iterative_ensemble_smoother/esmda.py:100]: Integer or a Generator I guess

File: src/iterative_ensemble_smoother/esmda.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+        # Only compute the covariance factorization once
+        # If it's a full matrix, we gain speedup by only computing cholesky once
+        # If it's a diagonal, we gain speedup by never having to compute cholesky
+        num_outputs = C_D.shape[0]
+
+        if isinstance(C_D, np.ndarray) and C_D.ndim == 2:
+            self.C_D_L = sp.linalg.cholesky(C_D, lower=False)
+        elif isinstance(C_D, np.ndarray) and C_D.ndim == 1:
+            assert len(C_D) == num_outputs
```
Comment (dafeda) [inline src/iterative_ensemble_smoother/esmda.py:134]: When would this be false?

**Review by dafeda — APPROVED** [review]
I've added a few comments but other than that it looks great.
Next step I guess is to test this API in ERT.

**Comment by dafeda** [comment]
Commits need squashing though 👍

**Pull request merged**